### ISSUE TITLE
Make it possible to re-map the device toggle buttons to use keys like the next and back ones on the side

### ISSUE THREAD

**Issue by carlspring** [pwr-Solaar/Solaar#812]
It would be great, if you could add functionality for re-mapping the "toggle device" on mice such as Logitech Anywhere MX2 to be the forward/next buttons on the left hand-side of the mouse. Perhaps it could use the next button to move on to the next numbered device and back to the previous one?

This has been bugging me for a while as a limitation of the Logitech Anywhere MX2, because I often have to switch between to laptops at the same time and turning the mouse over to press the button isn't very convenient and takes some time. Multiplying this by the number of times I have to do it a day, can really add up to a lot of wasted time. I don't use the side buttons I mentioned for anything and I'm sure there are many other people who would like to have this easier option for toggling.

It would be awesome, if thise feature is possible and can be implemented.

Many thanks in advance!

Martin

**Comment by pfps** [comment]
TLDR: Maybe

Solaar only changes the device, and doesn't act as a input device driver, so it won't modify the codes that the device sends to Linux.  However, it might be that Solaar can get your device to change which codes it produces in the way that you want.  I just put in a PR for Solaar to to this key/button reprogramming.  But if your device doesn't allow this key/button reprogramming then Solaar won't be able to help you.

So normally I would say to clone PR #810  and give that a try.  Unfortunately it's not going to be that simple, as it is extremely likely that the keys/buttons involved are unknown to Solaar so you would have to guess which unknown:#### is what you want to do.  I've requested more information on the key and button identifiers, but it might be a while before I get them.

**Comment by carlspring** [comment]
Hi,

Thank you for the quick reply! I will give it a go as soon as I can and I will report back.

**Comment by barolo** [comment]
Even Logitech Options don't allow remapping of this functionality, unfortunately, because switches for diff devices are on the underside of the mouse

**Comment by carlspring** [comment]
They've placed the device switch buttons in the most stupidest place, if you ask me... Every time you want to switch between different machines, you have to flip the mouse and switch to the next one. They could have added some buttons on the side, or something most convenient. They don't appear to have done much UX research. I think they thought that people would only really use it with one device at a time. I have machines + a phone. Flipping between them takes a lot of each day, as it adds up...

**Comment by barolo** [comment]
> They've placed the device switch buttons in the most stupidest place, if you ask me... Every time you want to switch between different machines, you have to flip the mouse and switch to the next one. They could have added some buttons on the side, or something most convenient. They don't appear to have done much UX research. I think they thought that people would only really use it with one device at a time. I have machines + a phone. Flipping between them takes a lot of each day, as it adds up...

Being the devil advocate, I'm guessing that they assume that you will use the "Flow" functionality for switching, physical switches solely as backup. AFAIK people do enjoy Flow. Unfortunately Linux systems are unsupported

**Comment by carlspring** [comment]
What is Flow?

**Comment by barolo** [comment]
> What is Flow?

It automatically switches device when you reach the screen border, merges screens pretty much, transparently. Amongst other things

**Comment by carlspring** [comment]
Yeah, but this probably requires software to be installed. The cool thing on corporate hardware in locked down environments is that typically Logitech mice work out of the box. You usually can't install third-party software on them. So, adding a Linux and Windows box to the picture and an Android phone would make it a nightmare to work with...

I wish they would just make their own Linux drivers or opensource them...

**Comment by barolo** [comment]
> Yeah, but this probably requires software to be installed. The cool thing on corporate hardware in locked down environments is that typically Logitech mice work out of the box. You usually can't install third-party software on them. So, adding a Linux and Windows box to the picture and an Android phone would make it a nightmare to work with...
> 
> I wish they would just make their own Linux drivers or opensource them...

YEah it needs Logitech Options, but for Linux you have this https://symless.com/synergy which predates it and works the same

**Comment by pfps** [comment]
Yes, I agree, out of scope for Solaar as far as I can tell.  I'm closing this issue until someone can figure out how Solaar might be used to solve the problem.

**Comment by pfps** [comment]
@barlo Thanks for checking out all these issues.   A big problem with adding features to Solaar is lack of access to different Logitech devices.  Having an active tester is very helpful.

**Comment by carlspring** [comment]
Logitech should be sponsoring this project, since they can't do things themselves! They should at least be providing some hardware and specs, (and I'd say even money to compensate you for all the reverse-engineering that needs to be done).

Thanks for letting me know. All the best!
